packages/solidjs/src/solidrouter.ts

```diff
@@ -0,0 +1,165 @@
+
+interface SolidRouterOptions {
+  useBeforeLeave: UserBeforeLeave;
+  useLocation: UseLocation;
+}
+
+function handleNavigation(location: string): void {
+  const client = getClient();
+  if (!client || !CLIENTS_WITH_INSTRUMENT_NAVIGATION.includes(client)) {
+    return;
+  }
+
+  startBrowserTracingNavigationSpan(client, {
+    name: location,
+    attributes: {
+      [SEMANTIC_ATTRIBUTE_SENTRY_OP]: 'navigation',
+      [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: 'auto.navigation.solidjs.solidrouter',
+      [SEMANTIC_ATTRIBUTE_SENTRY_SOURCE]: 'url',
+    },
+  });
+}
+
+function getActiveRootSpan(): Span | undefined {
+  const span = getActiveSpan();
+  return span ? getRootSpan(span) : undefined;
+}
+
+/** Pass-through component in case user didn't specify a root **/
+function SentryDefaultRoot(props: ParentProps): JSX.Element {
+  return props.children;
+}
+
+/**
+ * Unfortunately, we cannot use router hooks directly in the Router, so we
+ * need to wrap the `root` prop to instrument navigation.
+ */
+function withSentryRouterRoot(Root: Component<RouteSectionProps>): Component<RouteSectionProps> {
+  const SentryRouterRoot = (props: RouteSectionProps): JSX.Element => {
+    // TODO: This is a rudimentary first version of handling navigation spans
+    // It does not
+    // - use query params
+    // - parameterize the route
+
+    _useBeforeLeave(({ to }: BeforeLeaveEventArgs) => {
+      // `to` could be `-1` if the browser back-button was used
+      handleNavigation(to.toString());
+    });
+
+    const location = _useLocation();
+    createEffect(() => {
+      const name = location.pathname;
+      const rootSpan = getActiveRootSpan();
+
+      if (rootSpan) {
+        const { op, description } = spanToJSON(rootSpan);
+
+        // We only need to update navigation spans that have been created by
+        // a browser back-button navigation (stored as `-1` by solid router)
+        // everything else was already instrumented correctly in `useBeforeLeave`
+        if (op === 'navigation' && description === '-1') {
+          rootSpan.updateName(name);
+        }
+      }
+    });
+
+    return createComponent(Root, props);
+  };
+
+  return SentryRouterRoot;
+}
+
+/**
+ * A browser tracing integration that uses Solid Router to instrument navigations.
+ */
+export function solidRouterBrowserTracingIntegration(
+  options: Parameters<typeof browserTracingIntegration>[0] & SolidRouterOptions,
+): Integration {
+  const integration = browserTracingIntegration({
+    ...options,
+    instrumentPageLoad: false,
+    instrumentNavigation: false,
+  });
+
+  const { instrumentPageLoad = true, instrumentNavigation = true, useBeforeLeave, useLocation } = options;
+
+  return {
+    ...integration,
+    setup() {
+      _useBeforeLeave = useBeforeLeave;
+      _useLocation = useLocation;
+    },
+    afterAllSetup(client) {
+      integration.afterAllSetup(client);
+
+      const initPathName = WINDOW && WINDOW.location && WINDOW.location.pathname;
```

Comment:
m: Do we actually need this? Is this different from just letting the base `browserTracingIntegration` create the pageload span? If not, we can just pass `instrumentPageLoad` through (instead of forcing this to false) and only handle the navigation span in a custom way here?

Reply:
yeah, this is fine, this is really just for ourselves :)